oemof_b3/tools/data_processing.py

```diff
@@ -0,0 +1,165 @@
+
+    if np.all(df[index].array == df[index].array[0]):
+        value = df[index].array[0]
+        if value is None:
+            raise TypeError(
+                f"Your provided data is missing a {name}."
+                f"Please make sure you pass the {name} with {index}."
+            )
+        else:
+            return value
+    else:
+        raise ValueError(
+            f"The {name} of your provided data doesn't match for all entries. "
+            f"Please make sure to pass the {name} with {index}."
+        )
+
+
+def stack_timeseries(df):
+    """
+    This function stacks a Dataframe in a form where one series resides in one row
+
+    Parameters
+    ----------
+    df : pandas.DataFrame
+        DataFrame to be stacked
+
+    Returns
+    -------
+    df_stacked : pandas.DataFrame
+        Stacked DataFrame
+
+    """
+    _df = df.copy()
+
+    # Assert that _df has a timeindex
+    if not isinstance(_df.index, pd.DatetimeIndex):
+        raise TypeError(
+            "Your data should have a time series as an index of the format "
+            "'%Y-%m-%d %H:%M:%S'."
+        )
+
+    # Assert that frequency match for all time steps
+    if pd.infer_freq(_df.index) is None:
+        raise TypeError(
+            "No frequency of your provided data could be detected."
+            "Please provide a DataFrame with a specific frequency (eg. 'H' or 'T')."
+        )
+
+    _df_freq = pd.infer_freq(_df.index)
+    if _df.index.freqstr is None:
+        print(
+            f"User info: The frequency of your data is not specified in the DataFrame, "
+            f"but is of the following frequency alias: {_df_freq}. "
+            f"The frequency of your DataFrame is therefore automatically set to the "
+            f"frequency with this alias."
+        )
+        _df = _df.asfreq(_df_freq)
+
+    # Stack timeseries
+    df_stacked_cols = [
+        "var_name",
+        "timeindex_start",
+        "timeindex_stop",
+        "timeindex_resolution",
+        "series",
+    ]
+
+    df_stacked = pd.DataFrame(columns=df_stacked_cols)
+
+    timeindex_start = _df.index.values[0]
+    timeindex_stop = _df.index.values[-1]
+
+    for column in df.columns:
+        var_name = column
+        timeindex_resolution = _df[column].index.freqstr
+        series = [pd.Series(_df[column].values)]
+
+        column_data = [
+            var_name,
+            timeindex_start,
+            timeindex_stop,
+            timeindex_resolution,
+            series,
+        ]
+
+        dict_stacked_column = dict(zip(df_stacked_cols, column_data))
+        df_stacked_column = pd.DataFrame(data=dict_stacked_column)
+        df_stacked = df_stacked.append(df_stacked_column, ignore_index=True)
+
+    # Save name of the index in the unstacked DataFrame as name of the index of "timeindex_start"
+    # column of stacked DataFrame, so that it can be extracted from it when unstacked again.
+    df_stacked["timeindex_start"].index.name = _df.index.name
```

Comment:
@jnnr, do you see any difficulty implementing it like this for [this purpose](https://github.com/rl-institut/oemof-B3/pull/47#issuecomment-886609489)?

Reply:
I am not sure if this works. It seems that you assign the former index' name to the index name of df_stacked["timeindex_start"], which is a Series (or maybe a DataFrame with only one column). If you stack several timeseries with different index names, this does not work (not sure if this will ever happen, though).

Reply:
> If you stack several timeseries with different index names, this does not work (not sure if this will ever happen, though).

But the function can't stack several time series so far. Should we hence leave it like this for now or name it `timeindex` as default?

Reply:
It cannot, but if you could concat two stacked frames. I do not think this is too important, though. What we agreed on was the template, which is stacked. If you unstack from the template using your function, your timeindex will be called timeindex. This is reasonable, as the timeindex is specified in the stacked format by "timeindex_start" etc.

Reply:
Ok! Is this PR ready to be merged from your side?

Reply:
I am fine with merging :+1: